Accelerate hash table iterator with prefetching

NadavGigi opened the pull request:
This PR introduces improvements to the hashtable iterator, implementing prefetching technique described in the blog post [Unlock One Million RPS - Part 2](https://valkey.io/blog/unlock-one-million-rps-part2/) . The changes lay the groundwork for further enhancements in use cases involving iterators. Future PRs will build upon this foundation to improve performance and functionality in various iterator-dependent operations.

In the pursuit of maximizing iterator performance, I conducted a comprehensive series of experiments. My tests encompassed a wide range of approaches, including processing multiple bucket indices in parallel, prefetching the next bucket upon completion of the current one, and several other timing and quantity variations. Surprisingly, after rigorous testing and performance analysis, the simplest implementation presented in this PR consistently outperformed all other more complex strategies.

## Implementation
Each time we start iterating over a bucket, we prefetch data for future iterations:
- We prefetch the entries of the next bucket (if it exists).
- We prefetch the structure (but not the entries) of the bucket after the next.
This prefetching is done when we pick up a new bucket, increasing the chance that the data will be in cache by the time we need it.

## Performance


The data below was taken by conducting keys command on 64cores Graviton 3 Amazon EC2 instance with 50 mil keys in size of 100 bytes each. The results regarding the duration of “keys *” command was taken from “info all” command.
```
+--------------------+------------------+-----------------------------+
| prefetching        | Time (seconds)   | Keys Processed per Second   |
+--------------------+------------------+-----------------------------+
| No                 | 11.112279        | 4,499,529                   |
| Yes                | 3.141916         | 15,913,862                  |
+--------------------+------------------+-----------------------------+
Improvement:
Comparing the iterator without prefetching to the one with prefetching, 
we can see a speed improvement of 11.112279 / 3.141916 ≈ 3.54 times faster.
```


### Save command improvment

#### Setup:
- 64cores Graviton 3 Amazon EC2 instance.
-  50 mil keys in size of 100 bytes each.
-  Running valkey server over RAM file system.
-  crc checksum and comperssion off.

#### Results
```
+--------------------+------------------+-----------------------------+
| prefetching        | Time (seconds)   | Keys Processed per Second   |
+--------------------+------------------+-----------------------------+
| No                 | 28               | 1,785,700                   |
| Yes                | 19.6             | 2,550,000                   |
+--------------------+------------------+-----------------------------+
Improvement:
- Reduced SAVE time by 30% (8.4 seconds faster)
- Increased key processing rate by 42.8% (764,300 more keys/second)
```

NadavGigi commented on src/hashtable.c:
```diff
@@ -934,6 +946,76 @@
+}
+
+static void prefetchBucketEntries(bucket *b) {
+    for (uint8_t pos = 0; pos < numBucketPositions(b); pos++) {
```
Resolved

madolson commented:
How does this compare to having an iterator that actually returns a batch of items. Something like:
```
void *entries[7];
size_t num_entries;
entries = getBatchEntries(iterator, *num_entries);
if (entries) {
    for (size_t i = 0; i < num_entries; i++) {
        whatever(entries[i]);
    }
}
```

I generally prefer to avoid manually executing prefetching when we can just efficiently process the data, as we then give more hints to the compiler and the processor so it can efficiently do its own re-ordering and prefetching.

madolson commented:
> It's important to note that while we refer to 'threads' in this implementation, we're not actually using operating system threads.

Then don't name them threads, it makes the implementation much harder to follow.

NadavGigi commented:
> How does this compare to having an iterator that actually returns a batch of items. Something like:
> 
> ```
> void *entries[7];
> size_t num_entries;
> entries = getBatchEntries(iterator, *num_entries);
> if (entries) {
>     for (size_t i = 0; i < num_entries; i++) {
>         whatever(entries[i]);
>     }
> }
> ```
> 
> I generally prefer to avoid manually executing prefetching when we can just efficiently process the data, as we then give more hints to the compiler and the processor so it can efficiently do its own re-ordering and prefetching.

zuiderkwast commented:
I agree with Madelyn that the word "threads" is not not suitable. Can we call it batches?

This looks quite complex. Was this adapted from an implementation for dict?

For hashtable, there is a natural batch of one hashtable bucket (up to 7 elements). Can't we just prefetch all elements in a hashtable bucket and the child bucket if any? When we return the last element in the bucket, we'd prefetch the next bucket's elements. Then we just at read and return the elementa as before without keeping track of anything. It'd be much simpler.

For KEYS, we just want to return the elements so I guess it can be faster with a more complex solution, but for other usages like rdb dump and aof rewrite, we do more things after each call to hashtableNext so I would assume most of the benefit can be seen even with the simpler approach.

Did you try the simpler approach already?

uriyage commented on src/hashtable.c:
```diff
@@ -1828,77 +1902,114 @@ void hashtableReleaseIterator(hashtableIterator *iterator) {
+            info->state = BUCKET_PREFETCH_ENTRIES;
+            break;
+        default:
+            iter->num_of_active_buckets--;
```
why decrement if we are about to assert(0)?

NadavGigi replied:
Done.

uriyage commented on src/hashtable.c:
```diff
@@ -1828,77 +1902,114 @@ void hashtableReleaseIterator(hashtableIterator *iterator) {
+    }
+    /* When a bucket prefetch info attempts to pick up a new bucket
+     * and the iterator index exceeds the hashtable size,
+     * that bucket prefetch info finishes, decrementing num_of_active_buckets. */
```
Consider adding 
```c
iter->num_of_active_buckets = min(iter->num_of_active_buckets, iter->num_of_buckets);
```
 to avoid repeatedly calling `handleBucketState` for the same bucket when `iter->num_of_buckets is` < `iter->num_of_active_buckets`.

NadavGigi replied:
Its not a possible scenario. `iter->num_of_active_buckets` gets the value of `iter->num_of_bucket` during iterator init and then only decremented while we never changing `iter->num_of_bucket`.

NadavGigi commented:
> How does this compare to having an iterator that actually returns a batch of items. Something like:
> 
> ```
> void *entries[7];
> size_t num_entries;
> entries = getBatchEntries(iterator, *num_entries);
> if (entries) {
>     for (size_t i = 0; i < num_entries; i++) {
>         whatever(entries[i]);
>     }
> }
> ```
> 
> I generally prefer to avoid manually executing prefetching when we can just efficiently process the data, as we then give more hints to the compiler and the processor so it can efficiently do its own re-ordering and prefetching.

Thank you for your suggestion. If I understand correctly, this approach would necessitate changes to every piece of code that currently uses the iterator, not just the iterator implementation itself. However, I may have misinterpreted your proposal. Could you please provide further explanation or clarification on how this batch iterator could be implemented without requiring widespread changes to the existing codebase?


> > It's important to note that while we refer to 'threads' in this implementation, we're not actually using operating system threads.
> 
> Then don't name them threads, it makes the implementation much harder to follow.

Done.

NadavGigi commented:
> I agree with Madelyn that the word "threads" is not suitable. Can we call it batches?
> 
> This looks quite complex. Was this adapted from an implementation for dict?
> 
> For hashtable, there is a natural batch of one hashtable bucket (up to 7 elements). Can't we just prefetch all elements in a hashtable bucket and the child bucket if any? When we return the last element in the bucket, we'd prefetch the next bucket's elements. Then we just at read and return the elements as before without keeping track of anything. It'd be much simpler.
> 
> For KEYS, we just want to return the elements so I guess it can be faster with a more complex solution, but for other usages like rdb dump and aof rewrite, we do more things after each call to hashtableNext so I would assume most of the benefit can be seen even with the simpler approach.
> 
> Did you try the simpler approach already?

1. I have changed the name "threads" to bucketPrefetchInfo since its the number of buckets we scan in parrallel.
2. Yes, it was adapted from dict implementation.
3. No, I have tried a simpler version that benefits only from prefetching current bucket("batch 1" in performance table).
Following your comment, I have implemented the solution you suggested(indeed, very easy approach) and we can see the results here:
```
By conducting keys command on 64cores Graviton 3 Amazon EC2 instance with 50 mil keys in size of 100 bytes each.
+--------------------+------------------+-----------------------------+
| Implementation     | Time (seconds)   | Keys Processed per Second   |
+--------------------+------------------+-----------------------------+
| Iterator without   | 11.112279        | 4,499,529                   |
|    prefetching     |                  |                             |
| Viktor's approach  | 3.712602         | 13,467,954                  |
| batch iterator     | 3.336432         | 14,985,700                  |
+--------------------+------------------+-----------------------------+
```
***Batch iterator processes keys about 11% faster than Viktor's approach.***

zuiderkwast commented:
> _**Batch iterator processes keys about 11% faster than Viktor's approach.**_
> _**SAVE command time is the same for batch iterator and Viktor's approach.**_

Thanks for implementing and comparing this! From your table we can see that "my" approach is 3.0 times more throughput than without prefetching, compared to 3.33 times for your Nadav's approch.

Given the performance vs complexity, I'm leaning towards "my" simpler approach. Since you've already done the implementation, can you post it in a separate draft PR so we can see it?

Maybe the simple approach can be optimized a bit more, without too much increased complexity. (For example, prefetching the next 2-3 top-level buckets so we don't slow when there's an empty bucket.)

NadavGigi commented:
> > _**Batch iterator processes keys about 11% faster than Viktor's approach.**_
> > _**SAVE command time is the same for batch iterator and Viktor's approach.**_
> 
> Thanks for implementing and comparing this! From your table we can see that "my" approach is 3.0 times more throughput than without prefetching, compared to 3.33 times for your Nadav's approch.
> 
> Given the performance vs complexity, I'm leaning towards "my" simpler approach. Since you've already done the implementation, can you post it in a separate draft PR so we can see it?
> 
> Maybe the simple approach can be optimized a bit more, without too much increased complexity. (For example, prefetching the next 2-3 top-level buckets so we don't slow when there's an empty bucket.)

@zuiderkwast 
The implementation has been refined and now yields even better results. It's very similar to your approach but with some optimizations.

zuiderkwast commented on src/hashtable.c:
```diff
@@ -934,6 +934,32 @@ static inline incrementalFind *incrementalFindFromOpaque(hashtableIncrementalFin
+    }
+}
+
+static void prefetchNextFilledBucket(iter *iter, bucket *current_bucket) {
```
Please add a short comment above this and the other function.

For readers, it's useful to understand which memory is already in the cache (or assumed to be in the cache) and what is being prefetched. It makes it easier for readers to understand the optimization. Here, for example, it's assumed that `current_bucket` is already in the cache, right?

NadavGigi replied:
Done

zuiderkwast commented on src/hashtable.c:
```diff
@@ -934,6 +934,32 @@ static inline incrementalFind *incrementalFindFromOpaque(hashtableIncrementalFin
+
+static void prefetchNextFilledBucket(iter *iter, bucket *current_bucket) {
+    if (current_bucket->chained) {
+        prefetchBucketEntries(bucketNext(current_bucket));
```
Shouldn't we prefetch the next bucket (e.g. `valkey_prefetch(bucketNext(current_bucket))`) some time before we actually access the next bucket? The `prefetchBucketEntries` function accesses the memory of the bucket so it's a cold memory read if it hasn't been prefetched, isn't it?

NadavGigi replied:
Done.
Now we access only prefetched buckets.

zuiderkwast commented on src/hashtable.c:
```diff
@@ -934,6 +934,32 @@ static inline incrementalFind *incrementalFindFromOpaque(hashtableIncrementalFin
+        prefetchBucketEntries(bucketNext(current_bucket));
+    } else {
+        size_t table_size = numBuckets(iter->hashtable->bucket_exp[iter->table]);
+        for (size_t bucket_index = (size_t)iter->index + 1; bucket_index < table_size; bucket_index++) {
+            bucket *b = &iter->hashtable->tables[iter->table][bucket_index];
+            if (b->presence) {
+                prefetchBucketEntries(b);
```
The bucket at `b` has not yet been prefetched, so when we access `b->presence`, it is a cold memory lookup, isn't it? Or am I missing something?

Ideally, we should do `valkey_prefetch(b)` some time before we do `b->presence` and `prefetchBucketEntries(b)`.

NadavGigi replied:
The key improvement here is leveraging the CPU's automatic prefetching capabilities for contiguous memory:

1. For non-chained buckets in the table, we rely on the CPU to prefetch the empty buckets automatically, as they're in a contiguous block of memory.
2. We only explicitly prefetch the 'next of the next' bucket, not its entries.
3. For chained buckets, we still handle prefetching explicitly, as they may not be contiguous in memory.

zuiderkwast reviewed (commented):
Awesome! This implementation is much simpler and it's impressive that it's just as fast as the other one.

It seems that it could be optimized even a bit more to avoid a few more cache misses, or maybe am I missing something?

In general, the basic idea for prefetching is to only follow links one step at a time, with some time between each step, to make sure that we only access memory that has already been prefetched. I guess we can do this where we do `iter->index++`:

* Prefetch the bucket at index + 2 (just `valkey_prefetch(b)`)
* Prefetch the top-level elements and child-bucket pointer at index + 1. (The bucket itself has been prefetched the last time we incremented the index.)
* Prefetch the first level child-bucket elements at the current index. We start returning elements at the current top-level bucket elements that are already prefetched.

This makes sure that when we do index++ the next time, we only access memory that has already been previously prefetched. Right?

To also support long chains of child-buckets, we can prefetch the sub-child-bucket elements when we start returning elements from the child-bucket, and so on.

If this gets too complex and doesn't give any better performance, then ignore it.

zuiderkwast commented on src/hashtable.c:
```diff
@@ -934,6 +934,48 @@ static inline incrementalFind *incrementalFindFromOpaque(hashtableIncrementalFin
+    }
+}
+
+static bucket *getNextBucket(bucket *current_bucket, size_t bucket_index, hashtable *ht, int table_index) {
+    if (current_bucket->chained) {
+        return bucketNext(current_bucket);
```
This looks good. I see the names `getNextBucket` an `bucketNext` are very similar, but they are different. Maybe we can rename `bucketNext` to `getChildBucket` or something, to better distinguish child bucket from next bucket-index?

Also a comment for each function is always good. For this one, just mention that it returns the child bucket if any, otherwise the bucket at the next bucket index.

NadavGigi replied:
Done.

zuiderkwast replied:
Not done. I think you didn't read my comment properly. :)

I suggested to rename the other one, the one that returns the child bucket or null.

zuiderkwast replied:
Or something else, but to give them names that explain what they do and also make it clear that they are different and do different things.

NadavGigi replied:
Im sorry for the misunderstanding.
I have changed the name from `bucketNext` to `getChildBucket`.

zuiderkwast reviewed (approved):
This looks great to me now.

Can you run the benchmark again after the last changes?

NadavGigi commented:
> This looks great to me now.
> 
> Can you run the benchmark again after the last changes?

After running the benchmarks, I can confirm that the results remain the same as those presented in the table from the first comment.
